[v2]: CI Overhaul

JacksonBurns opened the pull request:
I'll update this once the PR is ready for review based on what I put in the commit messages, but in short - 
 - update the CI file
 - make the codebase pass the build, lint, and tests _on all platforms_

Let the fun begin...
[![image](https://github.com/chemprop/chemprop/assets/33505528/9ded9d57-ef25-4b7c-98f1-8e9e71a1b1eb)](https://www.google.com/url?sa=i&url=https%3A%2F%2Fgiphy.com%2Fgifs%2Fmfw-computer-dies-4no7ul3pa571e&psig=AOvVaw1QSP8nOtEuV1Ohfe8hs_72&ust=1709929143706000&source=images&cd=vfe&opi=89978449&ved=0CBIQjRxqFwoTCLiMr8v84oQDFQAAAAAdAAAAABAE)

JacksonBurns commented:
![image](https://github.com/chemprop/chemprop/assets/33505528/caf41d7c-4224-40e1-b03d-70ae86cdf145)

Has a nicely formatted action summary now.

JacksonBurns commented:
Have to table this for time being, but here are some notes from the future:
 - MacOS never gets to run the tests because the dependency install fails. It seems that conda doesn't have the binaries, pypi doesn't either, and the wheel building fails. My last effort is to change to compiler, but that probably won't work
 - On Python 3.12 all platforms fail the tests. When attempting to load a model from a checkpoint file, lightning somehow incorrectly tries to pass a `self` argument, which should have been filtered out by their own internal logic (they use inspect to get all of the model class's attributes and then filter out self) - I have no idea why this is broken, but lightning is famously finicky with loading from checkpoints so this might just require changing things seemingly randomly until it works.

JacksonBurns commented:
See https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/utilities/parsing.py#L51

JacksonBurns commented:
I have rolled back the commits attempting to fix the install problem.

JacksonBurns commented:
@davidegraff can you look at caf728d ? It's a (bad) initial attempt at removing `pytorch_lightning` in favor of native `torch` (but it's wrong because I don't know what I am doing).

JacksonBurns commented:
The good news is that all the tests are failing (meaning that they are all finally running!) :tada:

Apologies for the double ping, @davidegraff I know the `index` errors come from not properly passing the index tensor in the way that you showed in the issue, but I wasn't sure where to include it here.

JacksonBurns commented:
TODO: remove the torch_scatter installation workaround from the Docker container

davidegraff commented:
I don't feel that copying the code from `torch_scatter` is the right solution. We're not using the package for its python-level code, but rather because these operations are implemented efficiently at the C++/CUDA level. Copying the python source into the repo is the worst of all three options IMO because it's (1) slow and (2) not maintained (in that we're now responsible for code that we didn't actually write).

davidegraff commented:
This [reply](https://github.com/rusty1s/pytorch_scatter/issues/424#issuecomment-1975094408) suggests that installing from source works for MacOS

JacksonBurns commented:
> I don't feel that copying the code from `torch_scatter` is the right solution. We're not using the package for its python-level code, but rather because these operations are implemented efficiently at the C++/CUDA level. Copying the python source into the repo is the worst of all three options IMO because it's (1) slow and (2) not maintained (in that we're now responsible for code that we didn't actually write).

Agreed that we don't want to keep it exactly the same as the reference implementation. I've copied it here as (1) a starting point and (2) to establish the _structure_ I think we should shoot for when re-implementing this. I think we should have the scatter calls wrapped in a single util like I have done here - the actual implementation could be totally different, but this setup saves us having to update this in a lot of places if pytorch changes the API (since this feature is only in beta).

> This [reply](https://github.com/rusty1s/pytorch_scatter/issues/424#issuecomment-1975094408) suggests that installing from source works for MacOS, but I have no problem when I install: `pip install torch_scatter -f URL --no-index`

I tried the source install on MacOS and it didn't work on GitHub actions (failed during build). The former _might_ work, but the larger issue in my head is that this installation procedure (`install torch -> tomfoolery to install torch_scatter -> install chemprop`) has to go.

davidegraff commented:
But that still doesn't solve the problem of _slow_ python-level code though. I think it's much more preferable we move to a torch-native implementation based on a beta API than a slow one using yanked code for a few reasons:
1. The API is unlikely to change _that_ much considering that both `gather` and `scatter` (from which these APIs derive) are established APIs.
2. it's fast and "maintained" (I think it's a **_very_** bad idea to copy source code we don't understand for the core computation in our package)
3. we can limit the versions until the API is stable.

Also FWIW, I don't think this install process is too cumbersome. [It's the exact same ask as `pyg`](https://github.com/pyg-team/pytorch_geometric?tab=readme-ov-file#installation), a significantly larger package than ours. Consider that only until recently both RDKit and PyTorch required some finnicky install procedure and the community just dealt with it. I agree that we should strive for a package that is `pip`-installable, but the current procedure is really not that bad: it doesn't require specifying magic URLs, an involved setup script, etc.

And I'm not sure where the MacOS problems are coming from--I have no problems installing on my Mac:
```bash
$ conda create -n test_env python=3.11
$ conda activate test_env
$ pip install torch
$ pip install torch_scatter
$ python -c "import torch, torch_scatter; torch.manual_seed(42); print(torch_scatter.scatter_sum(torch.rand(10), torch.randint(4, size=(10,)), dim_size=4))"
tensor([0.6828, 1.9506, 0.5650, 1.6426])
$ pip uninstall torch torch_scatter -y
$ pip cache purge
$ pip install 'torch==2.2.0'
$ pip install torch_scatter
$ python -c "import torch, torch_scatter; torch.manual_seed(42); print(torch_scatter.scatter_sum(torch.rand(10), torch.randint(4, size=(10,)), dim_size=4))"
$ pip uninstall torch torch_scatter -y
$ pip cache purge
$ pip install torch-scatter -f https://data.pyg.org/whl/torch-2.2.0+cpu.html
$ python -c "import torch, torch_scatter; torch.manual_seed(42); print(torch_scatter.scatter_sum(torch.rand(10), torch.randint(4, size=(10,)), dim_size=4))"
```

JacksonBurns commented:
> But that still doesn't solve the problem of _slow_ python-level code though. I think it's much more preferable we move to a torch-native implementation based on a beta API than a slow one using yanked code for a few reasons:
> 
> 1. The API is unlikely to change _that_ much considering that both `gather` and `scatter` (from which these APIs derive) are established APIs.
> 2. it's fast and "maintained" (I think it's a **_very_** bad idea to copy source code we don't understand for the core computation in our package)
> 3. we can limit the versions until the API is stable.

We can change the actual implementation to be pure pytorch, I don't possess the skills to do so but that's a great idea. The only thing I am seeking to demonstrate here is the idea of keeping our existing calls to scatter in the codebase and keeping the 'nitty gritty' in one place. That 'nitty gritty' detail can look like whatever

> Also FWIW, I don't think this install process is too cumbersome. [It's the exact same ask as `pyg`](https://github.com/pyg-team/pytorch_geometric?tab=readme-ov-file#installation), a significantly larger package than ours. Consider that only until recently both RDKit and PyTorch required some finnicky install procedure and the community just dealt with it. I agree that we should strive for a package that is `pip`-installable, but the current procedure is really not that bad: it doesn't require specifying magic URLs, an involved setup script, etc.

It's not that bad, but it could be so much better :wink: 

> And I'm not sure where the MacOS problems are coming from--I have no problems installing on my Mac:It

It's not an issue on all Macs, but it doesn't even work all the time on Linux either - see [@kevingreenman's comment here](https://github.com/chemprop/chemprop/issues/580#issuecomment-1974952882).

davidegraff commented:
I'm fine with including black magic code and hiding it somewhere in the code-base if it's fast and something that we wrote. I just think it's _really_ bad practice to include it when it's not actually our code. Then we don't understand it but it's also technically our problem if it's bugged somewhere (as opposed to abstractly relying on its correctness via a 3rd-party package). As for a torch-native implementation, I included an experiment [here](https://github.com/chemprop/chemprop/issues/580#issuecomment-1922177577) that should provide some guidelines for switching the code over.

JacksonBurns commented:
I have pulled your suggestion for setting up the index and put it here:

https://github.com/chemprop/chemprop/pull/714/files#diff-81dbd7f93e533247fbfddddbf857f3974b4324778d27cd562fe072ebbb9d8f6dR43

and pulled from your suggested zero tensor initialization and put it here:

https://github.com/chemprop/chemprop/pull/714/files#diff-81dbd7f93e533247fbfddddbf857f3974b4324778d27cd562fe072ebbb9d8f6dR71

but as I have mentioned here: https://github.com/chemprop/chemprop/pull/714#issuecomment-1992736681

**This new code does not work because I don't know what I'm doing.**

The only reason I have even started to edit the scatter stuff is because chempropv2 is currently spuriously un-installable on half the world's computers because of our usage of `torch_scatter`. **I need more help getting this implementation to work** whether that's with pure pytorch throughout the codebase, custom python-based warppers in one place, or code shamelessly stolen from `torch_scatter`.

davidegraff commented:
my bad! I thought you had actually just copied most of the source from `torch_scatter` and placed it in inside our codebase. I see now what you're trying to do and that's on me, so I apologize for that. I can try and take a look at this later to see what's going on

davidegraff commented:
FWIW I don't think it's worth it to keep our `torch_scatter` client code the same and abstract away the "guts" to `utils/scatter.py` module. I think we should actually just change the client code to work with torch. The only spots we would need to change are seven uses across `nn/message_passing/base.py` and `nn/agg.py`.

I think these are all the recipes we would need: 

Starting from:
```python
>>> import torch_scatter
>>> X = torch.tensor([
    [0.3891, 0.0895, 0.3264, 0.5580],
    [0.4385, 0.3270, 0.6044, 0.6146],
    [0.7692, 0.8121, 0.4114, 0.0574],
    [0.8758, 0.6162, 0.7566, 0.2899],
    [0.7039, 0.5815, 0.0581, 0.2298],
    [0.0574, 0.8424, 0.0066, 0.7251],
    [0.5908, 0.7087, 0.5777, 0.2527],
    [0.8578, 0.8803, 0.8242, 0.3267],
    [0.2310, 0.4105, 0.9994, 0.6507],
    [0.8728, 0.1607, 0.9016, 0.1866]]
)
>>> index = torch.tensor([2, 3, 0, 1, 2, 0, 3, 3, 1, 0])
>>> index_torch = index.unsqueeze(1).repeat(1, X.shape[1])
>>> dim = 0
>>> dim_size = 4
```

### Sum
**`torch_scatter`**
```python
>>> torch_scatter.scatter(X, index, dim, dim_size=dim_size, reduce='sum')
tensor([[1.6994, 1.8152, 1.3196, 0.9690],
        [1.1068, 1.0266, 1.7560, 0.9405],
        [1.0930, 0.6710, 0.3845, 0.7878],
        [1.8871, 1.9160, 2.0063, 1.1941]])
```
**`torch`**
```python
>>> torch.empty(dim_size, X.shape[1]).scatter_reduce_(dim, index_torch, X, reduce='sum', include_self=False)
tensor([[1.6994, 1.8152, 1.3196, 0.9690],
        [1.1068, 1.0266, 1.7560, 0.9405],
        [1.0930, 0.6710, 0.3845, 0.7878],
        [1.8871, 1.9160, 2.0063, 1.1941]])
```

### Mean
**`torch_scatter`**
```python
>>> torch_scatter.scatter(X, index, dim, dim_size=dim_size, reduce='mean')
tensor([[0.5665, 0.6051, 0.4399, 0.3230],
        [0.5534, 0.5133, 0.8780, 0.4703],
        [0.5465, 0.3355, 0.1922, 0.3939],
        [0.6290, 0.6387, 0.6688, 0.3980]])
```
**`torch`**
```python
>>> torch.empty(dim_size, X.shape[1]).scatter_reduce_(dim, index_torch, X, reduce='mean', include_self=False)
tensor([[0.5665, 0.6051, 0.4399, 0.3230],
        [0.5534, 0.5133, 0.8780, 0.4703],
        [0.5465, 0.3355, 0.1922, 0.3939],
        [0.6290, 0.6387, 0.6688, 0.3980]])
```


### Softmax
**`torch_scatter`**
```
>>> torch_scatter.scatter_softmax(X, index, dim=0, dim_size=dim_size)
tensor([[0.4219, 0.3794, 0.5667, 0.5813],
        [0.2713, 0.2379, 0.3106, 0.4088],
        [0.3846, 0.3918, 0.3031, 0.2446],
        [0.6558, 0.5513, 0.4396, 0.4108],
        [0.5781, 0.6206, 0.4333, 0.4187],
        [0.1888, 0.4039, 0.2022, 0.4770],
        [0.3160, 0.3485, 0.3024, 0.2847],
        [0.4127, 0.4137, 0.3870, 0.3065],
        [0.3442, 0.4487, 0.5604, 0.5892],
        [0.4266, 0.2043, 0.4948, 0.2784]])
```
**`torch`**
> [!NOTE]
> This recipe uses _both_ `index` and `index_torch`. You can use `index_torch[:, 0]` in place of `index` 
```
>>> X_exp = X.exp()
>>> Z = torch.empty(dim_size, X.shape[1]).scatter_reduce_(dim, index_torch, X_exp, reduce='sum', include_self=False)
>>> X_exp / Z[index]
tensor([[0.4219, 0.3794, 0.5667, 0.5813],
        [0.2713, 0.2379, 0.3106, 0.4088],
        [0.3846, 0.3918, 0.3031, 0.2446],
        [0.6558, 0.5513, 0.4396, 0.4108],
        [0.5781, 0.6206, 0.4333, 0.4187],
        [0.1888, 0.4039, 0.2022, 0.4770],
        [0.3160, 0.3485, 0.3024, 0.2847],
        [0.4127, 0.4137, 0.3870, 0.3065],
        [0.3442, 0.4487, 0.5604, 0.5892],
        [0.4266, 0.2043, 0.4948, 0.2784]])
```

JacksonBurns commented:
@davidegraff thank you for the help! I have implemented the code you provided throughout the codebase and hopefully the tests pass. I may have incorrectly deduced the `dim_size` in some places, but we will find out soon enough.

davidegraff commented on chemprop/nn/message_passing/base.py:
```diff
@@ -245,7 +247,10 @@ def initialize(self, bmg: BatchMolGraph) -> Tensor:
 
     def message(self, H: Tensor, bmg: BatchMolGraph) -> Tensor:
-        M_all = scatter_sum(H, bmg.edge_index[1], 0)[bmg.edge_index[0]]
+        index_torch = bmg.edge_index[0].unsqueeze(1).repeat(1, H.shape[1])
```
I think this should be `edge_index[1]`

JacksonBurns commented:
@hwpang I think we're failing some of tests your helped with a lot, could you take a look at the CI failures? We are failing `test_roundtrip`, `test_checkpoint_is_valid`, and `test_checkpoint_roundtrip`

hwpang commented:
I see that the tests fail because one of the entries in the prediction tensor contains `nan`, while all the other entries are identical to the expected value. I quickly read through your changes, but it's not clear to me which part could introduce `nan`

davidegraff commented:
one possible guess is the use of `torch.empty` to initialize the destination tensors for aggregations with no input. That is, when you have a tensor `[0, 1, 2, 3, 4]` with aggregation indices `[0, 0, 1, 2, 1]` and a final size of `4`, the destination of tensor's 4th entry should be the identity (0 for sum, 1 for mul, etc.). `torch_scatter` handled this silently for us, but the recipes I shared above could possibly be the source of the issues.

A more concrete example where we aggregate into a tensor of size `4` but only use indices `1`, `2`, and `3`, leave `0` empty :
```python
>>> import torch_scatter
>>> X = torch.tensor([
    [0.3891, 0.0895, 0.3264, 0.5580],
    [0.4385, 0.3270, 0.6044, 0.6146],
    [0.7692, 0.8121, 0.4114, 0.0574],
    [0.8758, 0.6162, 0.7566, 0.2899],
    [0.7039, 0.5815, 0.0581, 0.2298],
    [0.0574, 0.8424, 0.0066, 0.7251],
    [0.5908, 0.7087, 0.5777, 0.2527],
    [0.8578, 0.8803, 0.8242, 0.3267],
    [0.2310, 0.4105, 0.9994, 0.6507],
    [0.8728, 0.1607, 0.9016, 0.1866]]
)
>>> index = torch.tensor([2, 3, 1, 1, 2, 2, 3, 3, 1, 2])
>>> index_torch = index.unsqueeze(1).repeat(1, X.shape[1])
>>> dim = 0
>>> dim_size = 4
```

If we take the `sum` or `mean` reduction of this tensor, then index `0` in the resulting destination tensor should be `0` (i.e., the additive identity), and **that's what we observe with `torch_scatter`**:

```python
>>> import torch_scatter
>>> torch_scatter.scatter(X, index, dim, dim_size=dim_size, reduce='sum')
tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [1.1642, 1.8691, 1.7626, 1.6657],
        [2.7350, 1.6438, 1.6975, 1.0318],
        [1.8871, 1.9160, 2.0063, 1.1940]])
```

We also get a seemingly correct answer using torch-native code:
```python
>>> torch.empty(dim_size, X.shape[1]).scatter_reduce_(dim, index_torch, X, reduce='mean', include_self=False)
tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [1.1642, 1.8691, 1.7626, 1.6657],
        [2.7350, 1.6438, 1.6975, 1.0318],
        [1.8871, 1.9160, 2.0063, 1.1940]])
```

But if we replace the destination from `torch.empty(dim_size, X.shape[1])` to `torch.ones(dim_size, X.shape[1])`, we get the wrong result **even though `include_self=False`**:
```python
>>> torch.ones(dim_size, X.shape[1]).scatter_reduce_(dim, index_torch, X, reduce='sum', include_self=False)
tensor([[1.0000, 1.0000, 1.0000, 1.0000],
        [1.1642, 1.8691, 1.7626, 1.6657],
        [2.7350, 1.6438, 1.6975, 1.0318],
        [1.8871, 1.9160, 2.0063, 1.1940]])
```

This indicates a problem with using `torch.empty` to create our destination tensors because we have no guarantee that the base-case entries are correct (i.e., entries where nothing gets aggregated into them--index `0` in this case).

### A possible fix
I think this should just imply we need to switch all `torch.empty` uses to `torch.zeros`, keeping in mind that if we ever use a multiplicative reduction in the future then we will need to use `torch.ones`

JacksonBurns commented:
@hwpang could you look at https://github.com/chemprop/chemprop/actions/runs/8329054142/job/22790506085?pr=714 around the pytest failures? They have to do with `load_from_checkpoint`

JacksonBurns commented:
Also thanks to @davidegraff , the previous test failures related to the scatter changes are fixed!

davidegraff commented:
sick, that seemed to fix it

hwpang commented:
@JacksonBurns that error looks similar this issue: https://github.com/Lightning-AI/pytorch-lightning/issues/18405. Strangest thing is that in both the 3.11 and 3.12 runs, we install the same version of pytorch lightning `pytorch_lightning-2.2.1-py3`

JacksonBurns commented:
@hwpang the issue mentions that implementing an `__init__` method in the sublass fixes the problem - does ours do that?

hwpang commented:
I'm not entirely sure what it meant to define an `__init__` function in our context. The `load_from_checkpoint` function is a class method under the `MulticomponentMPNN` class, which does have a `__init__` function.

JacksonBurns commented:
hm... that seems like probably the only way to 'fix' it. I have tried reverting the call to the superclass constructor call to the old syntax to see if it helps, but I doubt it.

This seems like some problem with lightning that we have no control over, so it may be best to just wrap the load_from_checkpoint method in MulticomponentMPNN and raise an error/warning about how Python 3.12 is bugged for that function and then wait for lightning to figure it out.

JacksonBurns commented:
> This seems like some problem with lightning that we have no control over, so it may be best to just wrap the load_from_checkpoint method in MulticomponentMPNN and raise an error/warning about how Python 3.12 is bugged for that function and then wait for lightning to figure it out.

@kevingreenman and @KnathanM there is an issue with loading `MulticomponentMPNN` models from checkpoints occurring only on Python 3.12, and the cause is somewhere upstream and very much out of our hands - would you be ok with adding a warning when loading and also skipping these tests if Python 3.12 is detected? The alternative would be to change `MulticomponentMPNN` to no longer inherit from `MPNN` and just copy-paste all of the stuff it needs, or else wait for the upstream to figure this out.

JacksonBurns commented:
[baaeca6](https://github.com/chemprop/chemprop/pull/714/commits/baaeca6b4f41585ae099c50d3063d688a0de28b2) disables the coverage checking on MacOS and Windows. There is a small chance that they might be responsible for the jobs hanging (although the issue reporting it later blames it on somethign unrelated), and not the out of memory error (or who knows, maybe it's _causing_ the OOM error). See https://github.com/nedbat/coveragepy/issues/1559

JacksonBurns commented:
the previous commit didn't fix it, but I'm now even more convinced that the coverage is a problem - see https://github.com/pytest-dev/pytest-cov/issues/599

JacksonBurns commented:
Turns out that wasn't the problem, but it did reveal that (for some unknown reason) Chemprop _itself_ actually hangs during the tests, not anything surrounding it :shrug: 

I am now going to go through the tedious process of just manually adding a pytest mark to each test that causes this issue and then skip said tests when running on the mac and windows runners. These tests will still be run on linux so it's not a big deal IMO

JacksonBurns commented:
note to self: the MPNN class also cannot be loaded from checkpoints because of the 'self' error that was breaking the MulticomponentMPNN

JacksonBurns commented:
This is a longshot, but previous versions of pytorch lightning would error out when attempting to pickle model that used lambda functions. We do that in chemprop to detect some module-level stuff so I'll try removing that and see if it somehow fixes this.

See here: https://pytorch-lightning.readthedocs.io/en/0.8.5/multi_gpu.html#make-models-pickleable

JacksonBurns commented:
New plan: wait and see if [46a1462](https://github.com/chemprop/chemprop/pull/714/commits/46a1462ed4275a17d7068825904897efbf9d7527) somehow fixes the runners on MacOS and Windows, if so then woohoo just need to disable checkpoint loading for the base network class on Python 3.12, if not then boo, still need to skip the failing tests and then do the checkpoint thing.

JacksonBurns commented:
This is the error we are seeing, and I believe we are already running these tests with one dataloader worker: https://pytorch.org/docs/stable/notes/windows.html#multiprocessing-error-broken-pipe

JacksonBurns commented:
Another related thread: https://lightning.ai/forums/t/crash-if-numworkers-0/2846/3

JacksonBurns commented:
Along the same lines as [46a1462](https://github.com/chemprop/chemprop/pull/714/commits/46a1462ed4275a17d7068825904897efbf9d7527) there may be something else in the codebase which cannot be pickled and is causing this error.

JacksonBurns commented:
Calling it quits for now, but there is a chance that the failure to load from checkpoint and the platform dependent failures are all due to the same cause - something in the codebase cannot be pickled.

JacksonBurns commented:
@KnathanM @hwpang @kevingreenman @davidegraff please see the PR description at the top for a summary of changes!

hwpang commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,114 @@
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+      # clone the repo, run black and autopep8 on it
```
Is this comment right? We are running flake8 not autopep8?

JacksonBurns replied:
good catch, we are running `flake8` which reports the errors that would be fixed by `autopep8`. I will fix this after the CI runs

hwpang commented on pyproject.toml:
```diff
@@ -34,14 +34,13 @@ dependencies = [
 [project.optional-dependencies]
 dev = ["black", "bumpversion", "flake8", "pytest", "pytest-cov"]
 docs = ["nbsphinx", "sphinx", "sphinx-argparse", "sphinx-autobuild", "sphinx-autoapi", "sphinxcontrib-bibtex", "sphinx-book-theme","nbsphinx-link","ipykernel"]
-test = ["parameterized > 0.8", "pytest >= 6.2", "pytest-cov"]
```
What did we use this for and why do we remove it?

JacksonBurns replied:
`parameterized` is used for running tests, but we don't use it anywhere (anymore, at least)

hwpang reviewed (commented):
Mostly LGTM with some questions

JacksonBurns commented:
TODO: review fixes
 - [ ] change the comment in the linting step of the CI to specify that we are running flake8 and not autopep8

JacksonBurns commented:
@KnathanM unfortunately creating the `_load_submodules` class was unsuccessful. I have left that commit on a separate branch here: bd710be85ec0e34855e1b7c57342aed43eeab13e but I have dropped it from this branch and instead pulled in the commit which disallow loading models outside of 3.11

davidegraff commented:
how much work would it be to cherry-pick commits removing `torch-scatter` from the codebase and move them into a separate PR? This PR is very big and would be easier to review as three or four separate PRs (i.e., one for each bullet in your review). At least 3/4 changes are consistent with addressing issues in the CI, but the `torch-scatter` change is more of a backend change that I'd like to review independently IMO.

JacksonBurns commented:
i dont think the torch-scatter change on its own won't pass the CI as it is on `v2/dev` rn, I'm keen to keep it all together

davidegraff commented:
yes but the CI fails for for a multitude reasons right now, each of them requiring a significant change to the codebase. I'm keen to review a change that significantly reworks a core piece of computation by itself.

JacksonBurns commented:
I've been pretty intentional about keeping each commit related to a distinct part of the overall change (like the scatter changes are all contained in 3d44bae), so rather than splitting it up I would recommend reviewing commit-by-commit.

JacksonBurns commented:
This PR is blocked by #741 which would allow us to drop d2ee130 and 749b2a9.

KnathanM commented:
> I've been pretty intentional about keeping each commit related to a distinct part of the overall change (like the scatter changes are all contained in [3d44bae](https://github.com/chemprop/chemprop/commit/3d44baec216e3934eae756f671ab927d50d35039)), so rather than splitting it up I would recommend reviewing commit-by-commit.

Is it possible then for me to review the torch scatter changes by making a throw away branch off of v2/dev and cherry-pick [3d44bae](https://github.com/chemprop/chemprop/commit/3d44baec216e3934eae756f671ab927d50d35039) onto it? Or does it depend on any of the previous commits?

JacksonBurns commented:
@KnathanM the changes will work and you can run tests locally when cherry picked, but the CI won't run tests because the formatting changes in previous commits are required to get to the test step. As long as that works?

KnathanM reviewed (commented):
I haven't looked at the torch scatter bits yet, so will cherry-pick that and run some tests. Some small comments on the rest of it. I am grateful you put so much into overhauling our CI for Chemprop.

JacksonBurns commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,118 @@
+# Continuous Integration for Chemprop - checks build, code formatting, and 
+# runs tests for all proposed changes and on a regular schedule
+#
+# Note: this file contains extensive inline documentation to aid with
+# knowledge transfer.
```
Personal preference, I think I was editing on a small screen at the time haha. Happy to change this if you like

KnathanM replied:
If you can change this comment and the comment that spans lines 3-4 to new line break after 100 characters, that would be good for me. Thanks.

JacksonBurns commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,118 @@
+      # clone the repo, attempt to build
+      - uses: actions/checkout@v4
+      - run: python -m pip install build
+      - run: python -m build .
```
The 'build' step refers to the literal python `build` command. If it fails we couldn't build the Python package at the end or install Chemprop to run the tests.

The test running doesn't use `build` but just installs the package. We could just go straight to running the tests, but it's kinda nice to have one expected point of failure for each step in the CI (like, the build only fails in one place, the tests only fail in one place, etc).

JacksonBurns commented on chemprop/data/datasets.py:
```diff
@@ -250,13 +250,13 @@ def normalize_inputs(
+                X = None if np.all(self.X_d is None) else self.X_d
             case "V_f":
-                X = None if np.all(self.V_fs == None) else np.concatenate(self.V_fs, axis=0)
+                X = None if np.all(self.V_fs is None) else np.concatenate(self.V_fs, axis=0)
             case "E_f":
-                X = None if np.all(self.E_fs == None) else np.concatenate(self.E_fs, axis=0)
+                X = None if np.all(self.E_fs is None) else np.concatenate(self.E_fs, axis=0)
             case "V_d":
-                X = None if np.all(self.V_ds == None) else np.concatenate(self.V_ds, axis=0)
+                X = None if np.all(self.V_ds is None) else np.concatenate(self.V_ds, axis=0)
```
This is a place where pep8 is totally unambiguous - _never_ compare `None` with anything other than `is`. I agree that we should just change this to check specific elements, but perhaps `all(i is None for i in self.V_ds)`?

KnathanM replied:
Yes that is fine and similar to what Hao-Wei does in hers.

JacksonBurns commented on chemprop/nn/predictors.py:
```diff
@@ -57,7 +67,8 @@ def train_step(self, Z: Tensor) -> Tensor:
 class _FFNPredictorBase(Predictor, HyperparametersMixin):
     """A :class:`_FFNPredictorBase` is the base class for all :class:`Predictor`\s that use an
-    underlying :class:`SimpleFFN` to map the learned fingerprint to the desired output."""
+    underlying :class:`SimpleFFN` to map the learned fingerprint to the desired output.
+    """
```
The linters don't complain about any of them, could be that the others don't have the parameters blocks so it doesn't care? Regardless if it passes `flake8` and `black` we don't need to worry.

JacksonBurns commented on tests/cli/test_cli_classification_mol.py:
```diff
@@ -148,3 +149,4 @@ def test_train_outputs(monkeypatch, data_path, tmp_path):
     checkpoint_path = tmp_path / "model_0" / "checkpoints" / "last.ckpt"
 
     model = MPNN.load_from_checkpoint(checkpoint_path)
+    assert model is not None
```
This is actually just to satisfy a `flake8` error. `model` would otherwise be an unused variable, so I put this in to satisfy that. The test could probably do something more involved? But I think this is fine.

KnathanM replied:
I think a better solution is to change it to not assign the output to a variable `MPNN.load_from_checkpoint(checkpoint_path)`. All the test is really doing is checking that we can load the model without errors, so we don't need to catch the output. But if we should add more tests and make it more involved, then it is probably better to leave `assert model is not None` as a place holder to remind us to add those tests. Your call.

JacksonBurns commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,118 @@
+      - name: Install dependencies
+        shell: bash -l {0}
+        run: |
+          python -m pip install flake8 pytest parameterized nbmake
```
Actually don't need flake8 either, thanks for pointing this out

KnathanM commented on chemprop/nn/agg.py:
```diff
@@ -112,6 +120,10 @@ def __init__(self, dim: int = 0, *args, output_size: int, **kwargs):
-        alphas = scatter_softmax(self.W(H), batch, self.dim)
-
-        return scatter(alphas * H, batch, self.dim, reduce="sum")
+        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
+        dim_size = batch.max().int() + 1
+        H_exp = H.exp()
+        Z = torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
+            self.dim, index_torch, H_exp, reduce="sum", include_self=False
+        )
+        return H_exp / Z[batch]
```
This isn't quite correct. Note how it doesn't make use of `self.W` while the current v2/dev version does. 
You can see a dimension mismatch error by replacing `agg = nn.MeanAggregation()` in `examples/training.ipynb` with `agg = nn.AttentiveAggregation(output_size=mp.output_dim)`.

Here is a version of forward that works by my tests.

```
def forward(self, H: Tensor, batch: Tensor) -> Tensor:
    dim_size = batch.max().int() + 1
    attention_logits = (self.W(H) - self.W(H).max()).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        self.dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        self.dim, index_torch, alphas * H, reduce="sum", include_self=False
    )
```

I tested it by replacing `agg = nn.AttentiveAggregation(output_size=mp.output_dim)` in `examples/training.ipynb` with:
```
import torch
from torch import Tensor
class NewAgg(nn.AttentiveAggregation):
    def forward(self, H: Tensor, batch: Tensor) -> Tensor:
        dim_size = batch.max().int() + 1
        attention_logits = (self.W(H) - self.W(H).max()).exp()
        Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
        )
        alphas = attention_logits / Z[batch]
        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
        return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, index_torch, alphas * H, reduce="sum", include_self=False
        )
agg = NewAgg(output_size=mp.output_dim)		
```

I'm note sure `attention_logits` is the best name for that variable. I'm not even sure that is a correct name, so open to suggestions.

Also note that I subtract `self.W(H).max()` before exponentiating. This doesn't change the output, but I think it helps with numerical stability? At least this is what [scatter_softmax](https://github.com/rusty1s/pytorch_scatter/blob/c095c62e4334fcd05e4ac3c4bb09d285960d6be6/torch_scatter/composite/softmax.py#L45) does. IMPORTANT: `torch_scatter` actually uses `scatter_max` to get the max from each index grouping (each molecule) in a batch while my version only get the single largest value across the whole `self.W(H)` vector. I couldn't find a way to get the max value for each grouping using only native pytorch. Using the max value of the whole vector may cause problems if one molecule's message passing output is much larger than the other molecules, so it would be good to find a more correct way to do this. 

Second IMPORTANT note, I found numerical instability when testing a tensor of all ones (`H = torch.ones_like(mp(bmg, V_d))`). The output should be a tensor of ones. Sometimes however some of the values are `1.0000001192092896` or `0.9999999403953552`. These are one bit above and below `1.0` (determined using [this method](https://stackoverflow.com/questions/16444726/binary-representation-of-float-in-python-bits-not-hex)). I also sometimes saw `0.9999992847442627`, `1.0000003576278687`, `1.0000004768371582`, `0.9999992251396179` and potentially others that are all off of `1.0` by a couple of bits. 

In the test below I simulate the output of message passing with a variable number of atoms, a batch size of 1 (so all molecule indices are 0, `batch = torch.zeros(n_atoms, dtype=torch.int64)` and an mp output of all ones (`H = torch.ones(n_atoms, mp_output_dim, dtype=torch.float32`). The dtypes are the same as what I get out of a `MolGraphDataLoader`. I compare three methods for the forward method of `AttentiveAggregation`. `forward_v2dev` is the current `torch_scatter` version. `forward_notscaled` is my version but without subtracting the max value before exponentiating. `forward_scaled` is the version I presented. The off-by-one-bit error doesn't happen for every random initialization of `W` (the attention layer that is just a `torch.nn.Linear(mp_output_dim, 1)`) so I loop through the test 1000 times to get a sense of how often it fails. My results are summarized below followed by the code.

| n_atoms | v2/dev errors | not scaled errors | scaled errors|
|---|---|---|---|
|3|0|326|0|
|4|0|0|0|
|5|0|193|0|
|6|0|442|0|
|7|0|466|0|
|8|0|447|0|
|9|0|375|0|
|10|1000|777|1000|
|11|1000|1000|1000|
|12|1000|1000|1000|
|13|0|432|0|
|14|1000|1000|1000|
|15|0|474|0|
|16|0|697|0|
|17|0|431|0|
|18|1000|1000|1000|
|19|1000|1000|1000|
|20|1000|832|1000|
|21|0|551|0|
|22|1000|585|1000|
|23|1000|1000|1000|
|24|1000|1000|1000|
|25|1000|1000|1000|
|26|1000|1000|1000|
|27|1000|1000|1000|
|28|1000|920|1000|
|29|1000|1000|1000|
|30|1000|1000|1000|
|31|1000|1000|1000|
|32|0|844|0|

I'll stop there, but I continued to see ups and downs in how often a couple bits are off. I'll note that most of the time `torch_scatter` and my scaled version have no failures while the unscaled version does, but there are also n_atoms values for which those two version fail everytime and the unscaled version only fails most of the time. 

Ultimately the fact that the `torch_scatter` version also has failures makes me think that we don't need to worry about this. And my proposed scaled version fails the same times that `torch_scatter` does so it could be a replacement. 

```
dim = 0
def forward_v2dev(H, batch, W):
    alphas = scatter_softmax(W(H), batch, dim)
    return scatter(alphas * H, batch, dim, reduce="sum")
def forward_notscaled(H, batch, W):
    dim_size = batch.max().int() + 1
    attention_logits = W(H).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        0, index_torch, alphas * H, reduce="sum", include_self=False
    )
def forward_scaled(H, batch, W):
    dim_size = batch.max().int() + 1
    attention_logits = (W(H) - W(H).max()).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        0, index_torch, alphas * H, reduce="sum", include_self=False
    )
```
```
n_atoms = 29

mp_output_dim = 1
fail_v2dev = 0
fail_notscaled = 0
fail_scaled = 0
n = 1000
for i in range(n):
    W = torch.nn.Linear(mp_output_dim, 1)
    batch = torch.zeros(n_atoms, dtype=torch.int64)
    H = torch.ones(n_atoms, mp_output_dim, dtype=torch.float32)
    v2dev = forward_v2dev(H, batch, W)
    notscaled = forward_notscaled(H, batch, W)
    scaled = forward_scaled(H, batch, W)
    if v2dev.item() != 1.0:
        fail_v2dev += 1
    if notscaled.item() != 1.0:
        fail_notscaled += 1
    if scaled.item() != 1.0:
        fail_scaled += 1
print(f"v2dev fails: {fail_v2dev}/{n}")
print(f"new fails: {fail_notscaled}/{n}")
print(f"fix fails: {fail_scaled}/{n}")
print(notscaled.item())
```

KnathanM commented:
I've reviewed the other torch_scatter changes and believe they are correct.

JacksonBurns commented:
With the merging of #741 I have dropped the commits related to disallowing loading models from checkpoint in Python 3.12 and rebased. I will now look at the review comment from Nathan and push another commit.

KnathanM reviewed (approved):
Thanks for all the work on this. I'll add a note for the future that we decided to use `self.W(H).exp()` instead of `(self.W(H) - self.W(H).max()).exp()` because it is unlikely that `self.W(H).exp()` will overflow a 32 bit floating point number.
